x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/support/Hasher.java

```diff
@@ -261,29 +421,108 @@ public static Hasher resolve(String name, Hasher defaultHasher) {
             case "clear_text":
                 return NOOP;
             default:
-                return defaultHasher;
+                throw new IllegalArgumentException("unknown hash function [" + name + "]");
         }
     }
 
-    public static Hasher resolve(String name) {
-        Hasher hasher = resolve(name, null);
-        if (hasher == null) {
-            throw new IllegalArgumentException("unknown hash function [" + name + "]");
+    public static Hasher resolveFromHash(char[] hash) {
+        String hashString = new String(hash);
+        if (hashString.startsWith(BCRYPT_PREFIX)) {
+            return Hasher.BCRYPT;
+        } else if (hashString.startsWith(PBKDF2_PREFIX)) {
+            return Hasher.PBKDF2;
+        } else if (hashString.startsWith(SHA1_PREFIX)) {
+            return Hasher.SHA1;
+        } else if (hashString.startsWith(MD5_PREFIX)) {
+            return Hasher.MD5;
+        } else if (hashString.startsWith(SSHA256_PREFIX)) {
+            return Hasher.SSHA256;
+        } else {
+            throw new IllegalArgumentException("unknown hash format for hash [" + hashString + "]");
+        }
+    }
+
+    public static boolean verifyHash(SecureString data, char[] hash) {
+        try {
+            final Hasher hasher = resolveFromHash(hash);
+            return hasher.verify(data, hash);
+        } catch (IllegalArgumentException e) {
+            // The password hash format is invalid, we're unable to verify password
+            return false;
         }
-        return hasher;
+    }
+
+    private static char[] getPbkdf2Hash(SecureString data, int cost) {
+        try {
+            // Base64 string length : ((4*n/3)+3) rounded up to the next multiple of 4 because of padding = 44
+            CharBuffer result = CharBuffer.allocate(PBKDF2_PREFIX.length() + String.valueOf(cost).length() + 2 + 32 + 44);
+            result.put(PBKDF2_PREFIX);
+            result.put(String.valueOf(cost));
+            result.put("$");
+            char[] salt = SaltProvider.salt(32);  // 32 characters for 64 bytes
```

Comment:
My recollection was that this salt provider was enough for our `ssha` implementation that is only used for the cache. I have concerns about it as a provider for stored hashes. The BCrypt implementation provides its own salt so we do not use this implementation there. I'd think it would be best to allocate a 64 byte array and fill the bytes with a securerandom. Then we can base64 encode those bytes as the salt. What do you think?

Reply:
I incidentally had this discussion with @talevy yesterday when I was telling him how the SaltProvider uses SecureRandom now in the process of this discussion we figured out that this change didn't make it to the PR somehow. 
I think the entropy of the generated salt will be the same in your approach and in the current one as both will generate a 32 character string from an alphabet of 64 characters. That said, in the spirit of removing code, I have no objections following your suggestion.

Reply:
Are you sure the entropy is the same? With the salt provider, the number of possible values for each position is 64. Meanwhile for actual bytes the value could be between 0-255 (256 possible values). I think the extra chars from base64 encoding is ok.

```java
SaltProvider.salt(32);  // 32 characters for 64 bytes
```

Given the value of the chars, each char fits in a single byte so we aren't really getting 64 bytes. (I am pretty sure about this, I didn't test though).

Reply:
> Given the value of the chars, each char fits in a single byte so we aren't really getting 64 bytes. (I am pretty sure about this, I didn't test though).

You're absolutely right, this tripped me. All chars used actually need 1 byte. 

> Are you sure the entropy is the same? 

Yes, but maybe I stated it incorrectly - or misused `entropy`.  My argument was that if by using either of the two approaches we end up with a char array of length 64 as a salt, so 512 bits, to exhaust that space one would need 2^512 attempts.